### ISSUE TITLE
[MNG-7693] NPE in createModelCache, modelCacheFactory is null

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8763]
**[Charles Oliver Nutter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=headius)** opened **[MNG-7693](https://issues.apache.org/jira/browse/MNG-7693?redirect=false)** and commented

This appears to be a new regression, possibly an incompatibility with polyglot-maven.

Maven 3.9.0 now produces the following error when building JRuby (https://github.com/jruby/jruby.git):

```
[] jruby $ mvn -Pbootstrap clean package
[INFO] Scanning for projects...
[ERROR] Internal error: java.lang.NullPointerException: Cannot invoke "org.apache.maven.repository.internal.ModelCacheFactory.createCache(org.eclipse.aether.RepositorySystemSession)" because "this.modelCacheFactory" is null -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.NullPointerException: Cannot invoke "org.apache.maven.repository.internal.ModelCacheFactory.createCache(org.eclipse.aether.RepositorySystemSession)" because "this.modelCacheFactory" is null
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:108)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:821)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:270)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.maven.repository.internal.ModelCacheFactory.createCache(org.eclipse.aether.RepositorySystemSession)" because "this.modelCacheFactory" is null
    at org.apache.maven.project.DefaultProjectBuilder.createModelCache (DefaultProjectBuilder.java:1002)
    at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:362)
    at org.sonatype.maven.polyglot.TeslaProjectBuilder.build (TeslaProjectBuilder.java:43)
    at org.apache.maven.graph.DefaultGraphBuilder.collectProjects (DefaultGraphBuilder.java:349)
    at org.apache.maven.graph.DefaultGraphBuilder.getProjectsForMavenReactor (DefaultGraphBuilder.java:340)
    at org.apache.maven.graph.DefaultGraphBuilder.build (DefaultGraphBuilder.java:76)
    at org.apache.maven.DefaultMaven.buildGraph (DefaultMaven.java:453)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:196)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:172)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:100)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:821)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:270)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/InternalErrorException
```

No additional information is given running with -X but I can provide that output if necessary.

This blocks JRuby CI due to the Windows Github Actions having updated to Maven 3.9.0 recently. Other platforms may soon also fail if updated.

Reproduce by using Maven 3.9.0 to build JRuby, which built correctly on previous Maven versions.

---

**Affects:** 3.9.0

**Issue Links:**
- [MNG-7704](https://issues.apache.org/jira/browse/MNG-7704) New field in DefaultProjectBuilder breaks maven-polyglot extension
 (_**"is duplicated by"**_)

**Remote Links:**
- [GitHub Pull Request #1008
](https://github.com/apache/maven/pull/1008)

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

[Charles Oliver Nutter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=headius) also, please look at two (one closed but just for posterity) related PRs:
* https://github.com/takari/polyglot-maven/pull/256 (closed but just FTR)
* https://github.com/takari/polyglot-maven/pull/257

Also, would be good if you'd chime in with JRuby bits into discussion. Thanks!

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Maven 4 is intentionally breaking compatibility in this area as it migrated from Plexus XML injection to JSR330 ctor injection, so there is even source level incompatibility in play...

**Comment by jira-importer** [comment]
**[Charles Oliver Nutter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=headius)** commented

Sorry for the late response, but I'm glad to see this has been addressed! I agree, making older clients still work is the best path forward, at least giving them some time to update to the new injection requirements.

We have mitigated this locally by using an older Maven release, but I look forward to getting rid of that restriction. Thank you!
